---
title: "Clustering and Differential Expression (10X 2.7k PBMCs)
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

## Goal

- To give you some experience visualizing genes in scRNA-Seq
- To teach you how to manke initial cell groupings.
- To learn how to select marker genes.

## Introduction

One of the first and most common tasks a computational biologist faces is the need to group cells meaningfully and to classify those groups of cells. This is one of the first analysis patterns that has become more or less a standard analysis pattern. Here we use Seurat to explore this analysis pattern. There are libraries that one could use in many of these steps to substitute other methodology. After your first steps here feel free to explore and extend this analysis.

```{r global_options, echo=FALSE}
knitr::opts_chunk$set(cache=TRUE)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80))
```

```{r}
library(Seurat)
library(dplyr)
library(gdata)
counts_matrix_filename = "shared_ro/pbmcs/pbmc3k.counts.matrix.gz"
myCountDF <- read.table(gzfile(counts_matrix_filename), header=T, row.names=1)
myCountMatrixSparse <- Matrix(as.matrix(myCountDF), sparse = TRUE)
seurat2obj = CreateSeuratObject(raw.data=myCountMatrixSparse)

pbmc.files <- "shared_ro/pbmcs/10X/hg19"
pbmc.data <- Read10X(data.dir = pbmc.files) # read in a 10X directory using Seurat
seurat2obj <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 3, min.genes = 200, project = "10X_PBMC")
```

We start with loading Seurat. Last time we loaded data into Seurat using a CellRanger output directory, this time we will create a Seurat object from a standard gzipped counts matrix. Often the hardest part of using R is getting data into the system; we want to given you experience in different types of input strategies.

*** Question ***
Was any filtering performed when we made the Seurat object? How would you find out about the behavior of the function?

## Examine contents of seurat2obj
```{r}
str(seurat2obj)
```

In the previous practical you learned a little about the slots in the Seurat object. You used a couple of them in your analysis. Some of the slots are automatically updated by Seurat as you move through analysis. Take a moment to look through the information, knowing the slots allow you to leverage work Seurat has already done for you. For example, how would you get the dimensions of the Seurat Objec tjust by using slots and not by using dim().

```{r}
VlnPlot(object = seurat2obj, features.plot = c("nGene"), group.by = c('orig.ident'))
```

Remember plotting violin plots, here we plot by what Seurat calls orig.ident. Identity is a concept that is used in the Seurat object ot refer to the cell identity (or cell groups). We will see how this updates as we go.

We will need some metadata to play with, let's try our old trick again and calculate and store percent mitochondrial reads.

```{r}
# let's just recompute percent mito again based on our filtered matrix:
mito.genes <- grep(pattern = "^mt-", x = rownames(x = seurat2obj@raw.data), ignore.case = TRUE, value = TRUE)
percent.mito <- Matrix::colSums(seurat2obj@raw.data[mito.genes, ]) /
                                        Matrix::colSums(seurat2obj@raw.data)

# AddMetaData adds columns to object@data.info, and is a great place to stash QC stats
seurat2obj <- AddMetaData(object = seurat2obj, 
                          metadata = percent.mito, 
                          col.name = "percent.mito")

VlnPlot(object = seurat2obj, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)
```

Let's not forget to filter!

```{r}
seurat2obj <- FilterCells(object = seurat2obj, subset.names = c("nGene", "percent.mito"), low.thresholds = c(200, -Inf), high.thresholds = c(2500, 0.05))
```

## Normalizing the data

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. 

```{r normalize}
seurat2obj <- NormalizeData(object = seurat2obj, 
                            normalization.method = "LogNormalize", 
                            scale.factor = 1e4)
```

Ok! Enough of a warm-up. On to new things!

## Detection of variable genes across the single cells

Seurat calculates highly variable genes and focuses on these for downstream analysis. **`FindVariableGenes`** calculates the average expression and dispersion for each gene, places these genes into bins, and then calculates a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko *et al*.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~2,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.

```{r var_genes}
seurat2obj <- FindVariableGenes(object = seurat2obj, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5, num.bin=20)  # if this fails, experiment with the num.bin setting
```

We can see the Seurat object slots have updated for the FindVariableGenes section. Let's use the slot to see how many variable genes we found.

```{r len_var_genes}
str(seurat2obj)
length(x = seurat2obj@var.genes)
```

We mentioned before that data needs special handling before performing PCA. Let's use ScaleData to scale and center the data and then perform PCA.

```{r pca_pre_regress}
seurat2obj <- ScaleData(object = seurat2obj)
seurat2obj <- RunPCA(object = seurat2obj, pc.genes = seurat2obj@var.genes, do.print = TRUE, pcs.print = 1:2, genes.print = 5, pcs.compute = 40, maxit = 500, weight.by.var = FALSE)
PCAPlot(object = seurat2obj, dim.1 = 1, dim.2 = 2)
```

There are a couple ways to visualize PCA in Seurate.

```{r}
# We can list top genes associated with components.
PrintPCA(seurat2obj)
# We can plot the top genes associated with components.
VizPCA(seurat2obj, pcs.use=1:2)
# We can plot a heat map of genes associated with each component (here component 1)
PCHeatmap(object = seurat2obj, pc.use = 1, cells.use = 500, do.balanced = TRUE, label.columns = FALSE)
```

Now that we have performed PCA we can plot metadata and genes on the cell groups. Here we use the FeaturePlot function to plot either on the PCA. We know we are using PCA (it is specified in reduction.use) specifically the first two PCA components (specified in dim.1 and dim.2).

```{r}
FeaturePlot(seurat2obj, dim.1=1, dim.2=2, reduction.use='pca', features.plot=c('nGene'))
```

How would you plot the same plot using component 2 and component 3. If we wanted to save the Seurate object we could use the save function (and later load with load(). We are not going to do this know but this is one of the way people share analysis, by saving and sharing Seurat R objects.

```{r saveobject}
# save(seurat2obj, file = "seurat2obj-pre_batch_correct.Robj")
```

PCA will create many components of variation but not all are useful. One way to select components is to use an elbow or scree plot. Let's plot one.

```{r}
PCElbowPlot(object = seurat2obj)
```

From this plot we are going to select the first 10 components to use in the t-SNE visualization.

```{r tsne}
seurat2obj <- RunTSNE(object = seurat2obj, dims.use = 1:10, do.fast = TRUE)
TSNEPlot(object = seurat2obj)
```

Let's plot complexity and percent mitochondrial reads on the t-SNE visualization. This will help visualize if either is driving the structure of the cell groupings.

```{r}
FeaturePlot(seurat2obj, reduction.use='tsne', features.plot=c("nGene","percent.mito"))
```

## Standard regressing out the nGene and percent.mito effects:

Sometimes there may be unwanted signal in the data that are associated with certain metadata. In these cases it may be helpful to regress out that signal. Here we practice with regressing out complexity and percent mitochondrial reads. We do this by including the variables to regress out in the ScaleData function (in the vars.to.regress parameter).

```{r}
# regress out the nGene effects
seurat2obj <- ScaleData(object = seurat2obj, vars.to.regress = c("nGene", "percent.mito"))
  
# rerun PCA on the regressed-out, 'cleaner' data
seurat2obj <- RunPCA(object = seurat2obj, pc.genes = seurat2obj@var.genes, 
                     do.print = FALSE, pcs.compute = 40, weight.by.var = FALSE)
  
# redo PCA and tSNE
seurat2obj <- RunTSNE(object = seurat2obj, 
                      dims.use = 1:10, # pca dimensions to use
                      seed.use = 12345, # random seed, layout will differ on diff settings
                      do.fast = TRUE) # runs Barnes-hut t-SNE
TSNEPlot(object = seurat2obj)
```

Now we can plot using the cleaned data. Let's do this with the original complexity and percent mitochondrial reads.

```{r}
# plot pca according to number of genes
FeaturePlot(seurat2obj, dim.1=1, dim.2=2, reduction.use='tsne',features.plot=c('nGene', 'percent.mito'))
```

## Cluster the cells

Note, most text in this section is lifted verbatim from the official Seurat2 tutorial. Modifications to code and/or text aim to make it generic for use in just about any target data set.

Seurat now includes a graph-based clustering approach compared to (Macosko *et al*.). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [[SNN-Cliq, Xu and Su, Bioinformatics, 2015]](http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract) and CyTOF data [[PhenoGraph, Levine *et al*., Cell, 2015]](http://www.ncbi.nlm.nih.gov/pubmed/26095251). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar gene expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'. As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard distance). To cluster the cells, we apply modularity optimization techniques [[SLM, Blondel *et al*., Journal of Statistical Mechanics]](http://dx.doi.org/10.1088/1742-5468/2008/10/P10008), to iteratively group cells together, with the goal of optimizing the standard modularity function.

The `FindClusters` function implements the procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.6-1.2 typically returns good results for single cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters are saved in the `object@ident` slot.

```{r cluster}
## Find clusters

# save.SNN = T saves the SNN so that the clustering algorithm 
#           can be rerun using the same graph
# but with a different resolution value (see docs for full details)
seurat2obj <- FindClusters(object = seurat2obj, reduction.type = "pca", 
                           dims.use = 1:10, resolution = 0.8, 
                           print.output = 0, save.SNN = TRUE)

# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = seurat2obj, do.label=T)
```

You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.
```{r}
# save(seurat2obj, file = "seurat2obj-post_batch_correct.Robj")
```

## Quality controling clusters

It is important to understand what is driving the clusters you are observing. This helps you to understand if you are looking at biological or technical effects. One of the easiest ways to do this is to plot technical metrics (like complexity, percent mictochondrial reads, or processing batches). We have already experimented with plotting metadata (or technical batches using the FeaturePlot()).

```{r}
### Plotting a prior known genes
FeaturePlot(object = seurat2obj, features.plot = c("CD14"), cols.use = c("grey", "blue"), reduction.use = "tsne")
```

Feel free to print other known genes before we calculate our markers.

### Assigning identity to clusters

Once you plot know marker genes, you may know the identity of some clusters. If so you can update the ident slot to new values.

```{r}
current.cluster.ids <- c(0, 1, 2, 3, 4, 5, 6, 7)
new.cluster.ids <- c("CD4 T cells", "CD14+ Monocytes", "B cells", "CD8 T cells", "FCGR3A+ Monocytes", "NK cells", "Dendritic cells", "Megakaryocytes")
seurat2obj@ident <- plyr::mapvalues(x = seurat2obj@ident, from = current.cluster.ids, to = new.cluster.ids)
TSNEPlot(object = seurat2obj, do.label = TRUE, pt.size = 0.5)
```

How did this work? We took the original cluster ids (originally by default numbered) and mapped them to new values using the mapvalues command.

## Selecting Marker Genes

```{r}
MIN_LOGFOLD_CHANGE = 1 # set to minimum required average log fold change in gene expression.
MIN_PCT_CELLS_EXPR_GENE = 0.1  # minimum percent of cells that must express gene in either clstr.

# Here we find all markers for 
all.markers = FindAllMarkers(seurat2obj,
                             min.pct = MIN_PCT_CELLS_EXPR_GENE,
                             thresh.use = MIN_LOGFOLD_CHANGE,
                             only.pos = TRUE,
                             test.use="bimod") # likelihood ratio test
```

Here we find markers for all clusters. There are other ways of selecting markers, feel free to read original Seurat tutorial for more details. Here we use the bimod method (Likelihood-ratio test). There are many other options including MAST.

Let's look at the top markers for different comparisons sorted by p-value.

```{r}
# sort all the markers by p-value
all.markers.sortedByPval = all.markers[order(all.markers$p_val),]

# take a look at the top most significant markers
head(all.markers.sortedByPval)
```

## Make a heatmap showing the top 10 markers

We can also make heat maps of the top markers.

```{r}
library(dplyr)
top10 <- all.markers.sortedByPval %>%  group_by(cluster)  %>% do(head(., n=10))
DoHeatmap(object = seurat2obj, genes.use = top10$gene, slim.col.label = TRUE, remove.key = TRUE)
```

Often one sees in publications plotting gene expression on t-SNE plot to see how specific the gene expression is to the clusters. Let's try a couple here.

```{r}
# examine the top 4 markers in the context of the tSNE plots:
FeaturePlot(seurat2obj, features.plot = all.markers.sortedByPval$gene[1:4])
```

Although you can always just perform differential expression on one cluster to get specific gene lists, if you performed all comparisons at once here is some advanced code to pull out unique marker genes to clusters.

## Get genes uniquely DE in each cluster:
```{r}
genes_uniquely_DE = all.markers.sortedByPval %>% dplyr::filter(avg_logFC > MIN_LOGFOLD_CHANGE) %>% group_by(gene) %>%  summarize(n=n()) %>%  filter(n==1)

genes_uniquely_DE.markers.sortedByPval =
  all.markers.sortedByPval[all.markers.sortedByPval$gene
                           %in% genes_uniquely_DE$gene,]

top_marker_each = genes_uniquely_DE.markers.sortedByPval %>%
      dplyr::filter(avg_logFC >= MIN_LOGFOLD_CHANGE) %>%
      dplyr::group_by(cluster) %>%
      do(head(., n=1))  # set to higher value to get that number of top markers each clst.
print(top_marker_each)
```

Here we write a for loop to iterate through each top marker and plot it using feature plot.

```{r}
for (gene in top_marker_each$gene) {
  FeaturePlot(seurat2obj, features.plot = gene)
}

```

## Other ways to explore marker genes:

We also do the same but plot violin plots instead of feature plots.

```{r}
for (i in 1:length(top_marker_each$gene)) {
  print(VlnPlot(seurat2obj, features.plot = top_marker_each$gene[i]))
}
```

## The DotPlot

Dotplots are a concise way of exploring percent of cells expressing a gene and the gene expression intensity. This give much of the same information a heatmap gives us without the resolution of looking at every cell. This can shield us from some of the noise caused by drop out and scales better with larger datasets.

```{r}
DotPlot(seurat2obj, genes.plot=unique(top_marker_each$gene))
```

Congratulations! You an plots cell groups and discover marker genes that describe cell groups. This is a very powerful analysis pattern often seen in publications. Well done!

## Sources

This practical is derived from the following resources, please visit them for updates and more details.

1. Seurat2:  http://satijalab.org/seurat
